streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java

```diff
@@ -223,9 +223,9 @@ private void emitNonJoinedOuterRecords(
             try (final KeyValueIterator<TimestampedKeyAndJoinSide<K>, LeftOrRightValue<V1, V2>> it = store.all()) {
                 TimestampedKeyAndJoinSide<K> prevKey = null;
 
+                boolean outerJoinLeftBreak = false;
+                boolean outerJoinRightBreak = false;
                 while (it.hasNext()) {
-                    boolean outerJoinLeftBreak = false;
-                    boolean outerJoinRightBreak = false;
```

Comment:
Ups... Wondering why we did not catch this in the unit tests?

Reply:
Currently, in the while-loop of the emitNonJoinedOuterRecords() method we are iterating over ALL the left- and right-side outerJoinRecords that are available in the outerjoin-store until we meet the **break;**.

The idea of the two outerJoinBreak flags was to keep track of when the window of the outerJoinRecords is not closed anymore, but this is only useful if the ordering of the KeyValueIterator is by timeStampedKey, and it is not:

The documentation says that the ordering of the KeyValueIterator is NOT guaranteed.

So, now I think we better can remove the outerJoinBreak flags and just check for each outerJoinRecord whether it belongs to a closed window or not, without any optimization. 
If the window has closed we can emit a nullJoinedValue. If the window is not closed yet we can continue with the next outerJoinRecord.

What do you think?  @mjsax @florin-akermann

Reply:
> Will try to make a unit-test for this.

Thanks! That's awesome!

> The documentation says that the ordering of the KeyValueIterator is NOT guaranteed.

But this is internal, right? This helper store is not plugable, so I think we could rely on ordering? -- We do store the data (ie, key) as `<timestamp><left/right-boolean><key>`, so we should be able to break the loop when we get the first timestamp which is too large (independent if it's a left or right record).

Reply:
Ok, that's good.

But we better not break the loop at the first timestamp which is too large.
For instance when the first timestamp which is too large belongs to a left record, there still can be a timestamp for a right record that is not yet too large.  This is because left and right records can have different (asymmetric, 0-sized) window sizes.

So, I would only break the loop when we have found the first timestamp that is too large for one-side record and then the second timestamp that is too large for the other-side record.

Unfortunately, this means that in the case of a left-join-call, there will only be found right-records in the outer join store and the loop will not break because it will try to find timestamps of left-records that are too large (that do not exist).

Unless you can find out that we are processing records for a left-join-call (and not an outer-join-call), we could break the loop inmediately at the first timestamp of a record which is too large.

What do you think?

Reply:
Good point about async window.

Right now, the `KStreamKStreamJoinProcessor` does not know if it's a left-outer or full-outer join. However, we know inside `KStreamImplJoin` which is the only place in which we create `KStreamKStreamJoin`, so it should be easy to pass in this information into the Processor :)

Reply:
Modified the emitNonJoinedOuterRecords-method for asymmetric windowing. 
Added a unit-test for asymmetric windows in a KStreamKStreamOuterJoinTest.java.
It fails if we break at the first timestamp that is too large.